hmmlearn/hmm.py

```diff
@@ -520,134 +525,294 @@ class GMMHMM(_BaseHMM):
+            elif self.covariance_type == 'full':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features, self.n_features))
+                self.covars_[:] = cv
+            elif self.covariance_type == 'diag':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features))
+                self.covars_[:] = np.diag(cv)
+            elif self.covariance_type == 'spherical':
+                self.covars_ = np.zeros((self.n_components, self.n_mix))
+                self.covars_[:] = cv.mean()
+
+    def _check(self):
+        super(GMMHMM, self)._check()
+
+        if not hasattr(self, "n_features"):
+            self.n_features = self.means_.shape[2]
+
+        # Checking covariance type
+        if self.covariance_type not in COVARIANCE_TYPES:
+            raise ValueError("covariance_type must be one of {0}"
+                             .format(COVARIANCE_TYPES))
+
+        self.weights_ = np.array(self.weights_)
+        # Checking mixture weights' shape
+        if self.weights_.shape != (self.n_components, self.n_mix):
+            raise ValueError("mixture weights must have shape "
+                             "(n_components, n_mix), "
+                             "actual shape: {}".format(self.weights_.shape))
+
+        # Checking mixture weights' mathematical correctness
+        if not np.allclose(np.sum(self.weights_, axis=1),
+                           np.ones(self.n_components)):
+            raise ValueError("mixture weights must sum up to 1")
+
+        # Checking means' shape
+        self.means_ = np.array(self.means_)
+        if self.means_.shape != (self.n_components, self.n_mix,
+                                 self.n_features):
+            raise ValueError("mixture means must have shape "
+                             "(n_components, n_mix, n_features), "
+                             "actual shape: {}".format(self.means_.shape))
+
+        # Checking covariances' shape
+        self.covars_ = np.array(self.covars_)
+        covars_shape = self.covars_.shape
+        needed_shapes = {
+            "spherical": (self.n_components, self.n_mix),
+            "tied": (self.n_components, self.n_features, self.n_features),
+            "diag": (self.n_components, self.n_mix, self.n_features),
+            "full": (self.n_components, self.n_mix,
+                     self.n_features, self.n_features)
+        }
+        needed_shape = needed_shapes[self.covariance_type]
+        if covars_shape != needed_shape:
+            raise ValueError("{!r} mixture covars must have shape {}, "
+                             "actual shape: {}"
+                             .format(self.covariance_type,
+                                     needed_shape, covars_shape))
+
+        # Checking covariances' mathematical correctness
+        from scipy import linalg
```

Comment:
To be honest, I did it, because similar function in scikit-learn does it. But it makes sense, this module is only used here. But I can do this way or another.